src/main/scala/com/microsoft/ml/spark/vw/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,421 @@
+
+  val cacheRows = new BooleanParam(this, "cacheRows", "Cache rows in multi-pass mode")
+  setDefault(cacheRows -> false)
+  def getCacheRows: Boolean = $(cacheRows)
+  def setCacheRows(value: Boolean): this.type = set(cacheRows, value)
+
+  val enableCacheFile = new BooleanParam(this, "enableCacheFile", "Enable local cache file")
+  setDefault(enableCacheFile -> false)
+  def getEnableCacheFile: Boolean = $(enableCacheFile)
+  def setEnableCacheFile(value: Boolean): this.type = set(enableCacheFile, value)
+
+  // abstract methods that implementors need to provide (mixed in through Classifier,...)
+  def getLabelCol: String
+  def getFeaturesCol: String
+
+  implicit class ParamStringBuilder(sb: StringBuilder) {
+    def appendParam[T](param: Param[T], option: String): StringBuilder = {
+      if (!get(param).isEmpty) {
+
+        param match {
+          case _: StringArrayParam => {
+            for (q <- get(param).get)
+              sb.append(' ').append(option).append(' ').append(q)
+          }
+          case _ => sb.append(' ').append(option).append(' ').append(get(param).get)
+        }
+      }
+
+      sb
+    }
+  }
+
+  val hashSeed = new IntParam(this, "hashSeed", "Seed used for hashing")
+  setDefault(hashSeed -> 0)
+
+  def getHashSeed: Int = $(hashSeed)
+  def setHashSeed(value: Int): this.type = set(hashSeed, value)
+
+  val numBits = new IntParam(this, "numBits", "Number of bits used")
+  setDefault(numBits -> 18)
+
+  def getNumBits: Int = $(numBits)
+  def setNumBits(value: Int): this.type = set(numBits, value)
+
+  private def createLabelSetter(df: DataFrame) = {
+    val labelColIdx = df.schema.fieldIndex(getLabelCol)
+
+    if (get(weightCol).isDefined) {
+      val weightColIdx = df.schema.fieldIndex(getWeightCol)
+      (row: Row, ex: VowpalWabbitExample) =>
+        ex.setLabel(row.getDouble(weightColIdx).toFloat, row.getDouble(labelColIdx).toFloat)
+    }
+    else
+      (row: Row, ex: VowpalWabbitExample) => ex.setLabel(row.getDouble(labelColIdx).toFloat)
+  }
+
+  /**
+    * Generate namespace info (hash, feature group, field index) for supplied columns.
+    * @param schema data frame schema to lookup column indices.
+    * @return
+    */
+  private def generateNamespaceInfos(schema: StructType): Array[NamespaceInfo] =
+    (Seq(getFeaturesCol) ++ getAdditionalFeatures)
+      .map(col => new NamespaceInfo(VowpalWabbitMurmur.hash(col, getHashSeed), col.charAt(0), schema.fieldIndex(col)))
+      .toArray
+
+  /**
+    * If multiple passes are requested and cache file is not enabled, cache the data in-memory.
+    * @param inputRows input data.
+    * @return iterator over the potentially cached data.
+    */
+  private def createDataIterator(inputRows: Iterator[Row], numPassesInSpark: Int) = {
+    if (numPassesInSpark == 1)
+    // avoid caching inputRows
+      () => inputRows
+    else {
+      // cache all row references to be able to quickly get the iterator again
+      // TODO: switch dataframe.cache()
```

Comment:
can you elaborate on the TODO?

Reply:
added this comment:

      // TODO: based on Mark's comment use Spark dataframe caching instead of materializing
      // row reference into memory